Publish explainer (#1)
WICG/lock-screen: README.md

# Lock Screen API Explainer

> Authors: Louise Brett, Glen Robertson

Reviewer: Please add some identifiers here, maybe just github usernames?

## Overview

This document proposes a Lock Screen API for enabling web apps on the lock screen of a device. This would allow users to use web apps for tasks such as checking a calendar, using a calculator, taking a photo, or taking a note, all without going through the extra step to unlock their device first. Such an API would be applicable to both desktop and mobile platforms, though it will require some operating system support to implement it.

When apps are run on the lock screen of a device, access to user data should be thoughtfully considered and restricted. For example a camera app running on the lock screen may let the user take photos but not view existing ones, and a calendar app on the lock screen may let the user see an upcoming event but not others. So the proposal includes isolation of lock screen apps, with methods to pass data to and from the lock screen to reduce the risk of apps accidentally exposing user data.

## Proposed API

> Apps that want to be shown on the lock screen would indicate this by declaring a URL in their manifest, to be used to launch the app on the lock screen. The URL must be within the _scope_ defined in the manifest. This URL could be a new manifest entry such as _lock\_screen\_start\_url_ but a more generic configuration would allow other custom launch actions to be added in future. We propose adding a _special\_launchers_ list to the manifest, in which items declare a _URL_ and a _purpose_ that may be recognised by user agents.

Reviewer: This is now outdated. Consider:
```suggestion
Apps that want to be shown on the lock screen would indicate this by declaring a URL in their manifest, to be used to launch the app on the lock screen. The URL must be [within the scope](https://www.w3.org/TR/appmanifest/#dfn-within-scope) defined in the manifest. This URL could be a new manifest entry such as `lock_screen_start_url` or a more-extensible `lock_screen` dict with a `start_url` member (proposed).
```
(also replace italics with codeblocks for code)

```
"lock_screen": {
  "start_url": "/some/url" 
}
```

To use the API users would start by calling window.getLockScreenData().

```js

> const data = await window.getLockScreenData();

Reviewer: IMO this example fits better if placed right after "The app may use this API to send data from the lock screen instance to the unlocked instance of the app and vice versa" in the paragraph below, or after that paragraph.

> await data.setData(key, content);

Reviewer: Maybe add some dummy strings here "my-key" "my-content"?

```

### Passing data between lock screen instance and unlocked instance

Apps running on the lock screen must be isolated from regular user data such as cookies and local storage. This means there will be two instances of the app: the main instance that is used when the device is unlocked, and a separate instance that can be launched on the lock screen without access to user data. The app may use this API to send data from the lock screen instance to the unlocked instance of the app and vice versa. Ideally, the web app running on the lock screen should not ask the user to authenticate, operating in an anonymous mode and using the lock screen API to sync data to the user's profile. The goal is to avoid accidentally exposing user data to the lock screen by forcing the app to explicitly send any data needed via this API.

### Notifying apps about available data

> When the device is unlocked, if there is data available, the instance of the app running in the unlocked context is notified by listening for the _dataitemsavailable_ service worker event.

Reviewer: Use backticks instead of underscores for code blocks
